presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/query/PinotQueryGenerator.java

```diff
@@ -0,0 +1,408 @@
+        final String pql;
+        final List<Integer> columnIndicesExpected;
+        final int numGroupByClauses;
+        final boolean haveFilter;
+        final boolean isQueryShort;
+
+        @JsonCreator
+        public GeneratedPql(@JsonProperty("table") String table, @JsonProperty("pql") String pql, @JsonProperty("columnIndicesExpected") List<Integer> columnIndicesExpected, @JsonProperty("numGroupByClauses") int numGroupByClauses, @JsonProperty("haveFilter") boolean haveFilter, @JsonProperty("isQueryShort") boolean isQueryShort)
+        {
+            this.table = table;
+            this.pql = pql;
+            this.columnIndicesExpected = columnIndicesExpected;
+            this.numGroupByClauses = numGroupByClauses;
+            this.haveFilter = haveFilter;
+            this.isQueryShort = isQueryShort;
+        }
+
+        @JsonProperty("pql")
+        public String getPql()
+        {
+            return pql;
+        }
+
+        @JsonProperty("columnIndicesExpected")
+        public List<Integer> getColumnIndicesExpected()
+        {
+            return columnIndicesExpected;
+        }
+
+        @JsonProperty("numGroupByClauses")
+        public int getNumGroupByClauses()
+        {
+            return numGroupByClauses;
+        }
+
+        @JsonProperty("table")
+        public String getTable()
+        {
+            return table;
+        }
+
+        @JsonProperty("haveFilter")
+        public boolean isHaveFilter()
+        {
+            return haveFilter;
+        }
+
+        @JsonProperty("isQueryShort")
+        public boolean isQueryShort()
+        {
+            return isQueryShort;
+        }
+
+        @Override
+        public String toString()
+        {
+            return toStringHelper(this)
+                    .add("pql", pql)
+                    .add("table", table)
+                    .add("columnIndicesExpected", columnIndicesExpected)
+                    .add("numGroupByClauses", numGroupByClauses)
+                    .add("haveFilter", haveFilter)
+                    .add("isQueryShort", isQueryShort)
+                    .toString();
+        }
+    }
+
+    class PinotQueryPlanVisitor
+            extends PlanVisitor<PinotQueryGeneratorContext, PinotQueryGeneratorContext>
+    {
+        private final ConnectorSession session;
+        private final boolean preferBrokerQueries;
+
+        protected PinotQueryPlanVisitor(ConnectorSession session, boolean preferBrokerQueries)
+        {
+            this.session = session;
+            this.preferBrokerQueries = preferBrokerQueries;
+        }
+
+        @Override
+        public PinotQueryGeneratorContext visitPlan(PlanNode node, PinotQueryGeneratorContext context)
+        {
+            throw new PinotException(PINOT_UNSUPPORTED_EXPRESSION, Optional.empty(), "Don't know how to handle plan node of type " + node);
+        }
+
+        protected VariableReferenceExpression getVariableReference(RowExpression expression)
```

Comment:
this function should be private static and only belongs to `PinotQueryGenerator` instead of the visitor